### PR TITLE
Debian: fix overwriting of config settings on upgrade

### LINKED ISSUE #4440: debian postinst craps over existing server_name setting
If `server_name.yaml` is edited manually, the postinst overwrites it on upgrade. A few points:
* if it's autogenerated, the file should say so
* an upgrade shouldn't touch it anyway (the postinst can tell if it's an upgrade or a fresh install)
* even on fresh install, if the file exists, maybe we should leave it alone?

Update 2019/01/18: likewise, `report_stats.yaml` is also overwritten according to the debconf database. (In some ways this is worse, because it may be some time before the admin notices it.)

### PR DESCRIPTION
**Pull request by richvdh**
Make sure that users' changes to the config files are preserved.

Fixes #4440.

### CODE REVIEW

File: debian/manage_debconf.pl [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+} elsif ($cmd eq 'read') {
+    $read = 1;
+} elsif ($cmd eq 'update') {
+    $update = 1;
+} else {
+    die "unknown command '$cmd'\n";
+}
+
+my @files = (
+    "homeserver.fred.yaml",
```
Comment (hawkowl) [inline debian/manage_debconf.pl]: ... `.fred`?

Comment (richvdh) [reply debian/manage_debconf.pl]: oops. good spot, thanks.

**Comment by richvdh** [comment]
FTR: The CI is failing because I haven't added a newsfragment (it's in the debian changelog instead)

**Comment by erikjohnston** [comment]
I wonder whether an easier approach would be to only update the files on a) initial install and b) when a user changes the debconf stuff manually (e.g. via dpkg-reconfigure)?

**Comment by erikjohnston** [comment]
Trying to manipulate yaml with regex makes me very sad tbh, can we not do:

1. Read config vars via `python -m synapse.config read <var_nam>`
2. Completely clobber the auto generated config file

**Comment by richvdh** [comment]
> 1\. python -m synapse.config read <var_nam>

the trouble with that is that it will fail if someone has any sort of error in their config file (nb: not just a yaml error - anything that we would reject as an invalid synapse config). I fear this will leave us back in the same place where people randomly get their configs overwritten

So I considered writing a new python module to read the raw yaml and fish out the required config, but it felt a bit like overkill (plus I'd rather we were tolerant of yaml formatting errors, for the same reason as above).

> 2\. Completely clobber the auto generated config file

I guess, but I'd rather we preserve comments, extra configs that people choose to throw in, a decision to move the autogenerated settings to different files, etc.

**Comment by erikjohnston** [comment]
My main concern here is that its just going to subtly break because someone has done something weird in the config file, like using the valid yaml syntax of:

```yaml
server_name: |-
   example.com
```

> I guess, but I'd rather we preserve comments, extra configs that people choose to throw in, a decision to move the autogenerated settings to different files, etc.

TBH, a lot of packages don't allow any of that stuff, and people don't complain so long as there is a big screaming warning at the top.

**Comment by richvdh** [comment]
I don't think the case you point out will cause any breakage, except that the debconf ui will show the server name as `|-`. Agreed if they then change it, it will break, but ho hum

**Pull request merged**